search.py

```diff
@@ -109,8 +121,29 @@ def cosine_score(tokens_arr):
             continue
 
         # 2. Obtain the second vector's (query vector's) value for pointwise multiplication
-        # Calculate the weight entry of the term in the query, of the term-document matrix
-        query_term_weight = get_query_weight(posting_list.size, term_frequencies[term])
+
+        query_term_weight = get_query_weight(posting_list.size, term_frequencies[term]) # without Rocchio
+
+        if (True):
+            # Apply Rocchio Algorithm for Query Refinement:
+            # ie the weight entry of the term in the refined (aka finalised) query in the term-document matrix
+            # Note we treat both the initial query and relevant-marked documents as documents
+
+            # this is the initial query's contribution to tf to calculate refined query value for this term
+            initial_query_value = term_frequencies[term]
+
+            # calculate the centroid value for tf for calculating refined query value for this term (done later)
+            accumulated_value = 0
+            for doc_id in relevant_docids:
+                # We have the posting list for the term, we just need to scan through it
+                # to get the contribution of this current doc_id marked relevant
+                accumulated_value += find_term_specific_weight_for_specified_id(doc_id, posting_list)
```

Comment:
`find_term_specific_weight_for_specified_id` -> `get_document_vector`

Reply:
technically not true, we only get a particular entry for that vector, which is represented by its weight